### PR TITLE
ThreadSanitizer support for arm64

### PR DESCRIPTION
**Pull request by ghost**
This adds TSan support to Linux/arm64. FreeBSD would probably work as well once TSan is available on FreeBSD/arm64 (on FreeBSD, TSan is only available on amd64 and powerpc64 at the moment); I have no idea regarding macOS (and no such hardware available to tinker on).

All the tests pass, except for `tests/effects/evenodd.ml` which triggers the OOM killer on the 8GB machine I tested on (but passes with the iteration count lowered from `100_000` to `80_000`).

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1692,8 +1693,20 @@ AS_IF([test "x$enable_tsan" = "xyes" ],
...
           [AC_MSG_ERROR([thread sanitizer not supported with compiler \
-$cc_basename"])]
+$ocaml_cv_cc_vendor"])]
          )],
+        [AC_MSG_ERROR([thread sanitizer not supported on system $system])]
+       )],
+    [arm64],
+      [AS_CASE(["$system"],
+        [linux],
+        [AS_CASE(["ocaml_cv_cc_vendorcc_basename"],
```
Comment (MisterDA) [inline configure.ac]: copy-paste typo?

Comment (ghost) [reply configure.ac]: Oops. Last-minute edit botch under vi, sorry. Thanks for noticing!

**Comment by tmcgilchrist** [comment]
Testing this on macOS/arm64 (M2 with MacOS Sonoma) with https://github.com/ocaml/ocaml/commit/f8d0927066700a1e55428190a9e5b8eb437d5c87 passes the compiler test suite. Are there other tests that need to be run?

File: asmcomp/arm64/emit.mlp [hunk 2]
```diff
@@ -962,9 +959,11 @@ let emit_instr env i =
         let offset = Domainstate.(idx_of_field Domain_dls_root) * 8 in
         `	ldr	{emit_reg i.res.(0)}, [{emit_reg reg_domain_state_ptr}, {emit_int offset}]\n`
     | Lop(Ireturn_addr) ->
-        invalid_arg (
-          "Support for Ireturn_addr is not implemented on architecture "
-          ^ Config.architecture)
+        let n = frame_size env in
```
Comment (tmcgilchrist) [inline asmcomp/arm64/emit.mlp:962]: The x86 version of this is clearer, naming this value offset and doing the calculation on a single line.

https://github.com/ocaml/ocaml/blob/7ceab32dd9a759253059d48cc5facb38211a4be0/asmcomp/amd64/emit.mlp#L743-L744

File: runtime/arm64.S [hunk 3]
```diff
@@ -663,6 +829,23 @@ FUNCTION(caml_reraise_exn)
...
+/* When TSan support is enabled, this routine should be called just before
+   raising an exception. It calls __tsan_func_exit for every OCaml frame about
+   to be exited due to the exception.
+   Takes no arguments, clobbers C_ARG_1, C_ARG_2, C_ARG_3 and potentially all
+   caller-saved registers of the C calling convention. */
+FUNCTION(caml_tsan_exit_on_raise_asm)
+        CFI_STARTPROC
+        mov     x0, x30        /* arg1: pc of raise */
+        mov     x1, sp         /* arg2: sp of raise */
+        mov     x2, TRAP_PTR   /* arg3: sp of handler */
```
Comment (tmcgilchrist) [inline runtime/arm64.S:842]: The comment above mentions `C_ARG_1` etc getting clobbered but the mov instructions here don't reference those names. To be consistent within this file it would be good to use them where the comments mention those symbols. 

This seems to be the case with many of the new TSan sections.

**Comment by gasche** [comment]
As someone who is unfamiliar with amr64, I am struck by the abundance of the following push/pop patterns:

Two registers:
```
        stp     x0, x1, [sp, -16]!
        CFI_ADJUST(16)
[...]
        ldp     d0, d1, [sp], 16
        CFI_ADJUST(-16)
```

and

One register:
```
        str     x0, [sp, -16]!
        CFI_ADJUST(16)
[...]
        ldr     x0, [sp], 16
        CFI_ADJUST(-16)
```

This makes it tempting to define `{PUSH,POP}{1,2,3,4,5,6}` macros, to name those common patterns. But maybe this is just music to the ears of arm64 programmers and there is no need to hide it under a small abstraction layer.

**Comment by ghost** [comment]
> Testing this on macOS/arm64 (M2 with MacOS Sonoma) with [f8d0927](https://github.com/ocaml/ocaml/commit/f8d0927066700a1e55428190a9e5b8eb437d5c87) passes the compiler test suite. Are there other tests that need to be run?

No, the compiler testsuite is a good exercizer already. Thanks for testing!

**Comment by ghost** [comment]
> As someone who is unfamiliar with amr64, I am struck by the abundance of the following push/pop patterns:
> [...]
> This makes it tempting to define `{PUSH,POP}{1,2,3,4,5,6}` macros, to name those common patterns. But maybe this is just music to the ears of arm64 programmers and there is no need to hide it under a small abstraction layer.

I think that having PUSH/POP macros would end up being confusing to seasoned arm programmers, although it might make some sense from a code compactness point of view (and also would make sure there is no risk of forgetting the CFI updates). But I am obviously a bit biaised in my opinion here, it might be interesting to know what other contributors to the arm64 back-end think of this.

**Comment by gasche** [comment]
This was purely a side-comment, feel free to drop the suggestion if knowledgeable people don't need it.

**Comment by gasche** [comment]
(cc @OlivierNicole @fabbing)

File: configure.ac [hunk 4]
```diff
@@ -1702,7 +1714,7 @@ AS_IF([$tsan],
   LDFLAGS="$SAVED_LDFLAGS"
 
   AS_CASE([$ocaml_cv_cc_vendor],
-    [gcc-[[0123456789]]-*|gcc-10-*|clang-*],
```
Comment (tmcgilchrist) [inline configure.ac]: It's not clear why this change is necessary, could you comment why it is needed?

File: runtime/arm64.S [hunk 5]
```diff
@@ -898,11 +1216,14 @@ L(frame_runstack):
...
-        mov     x0, x20
-        mov     x1, x19
+    /* switch directly to parent stack */
         mov     sp, x21
         CFI_RESTORE_STATE
+    /* Signal to TSan that we exit caml_runstack (no registers to save here) */
+        TSAN_EXIT_FUNCTION
+    /* pick correct return value */
+        mov     x0, x20
+        mov     x1, x19
```
Comment (tmcgilchrist) [inline runtime/arm64.S:1226]: Does the re-ordering here of moving the return value loading impact the CFI_RESTORE_STATE directive? It seems correct with respect to the TSAN_EXIT_FUNCTION behaviour here and in `amd64.S`

File: configure.ac (hunk 4 above)
Comment (MisterDA) [reply configure.ac]: I suppose we don't expect [GCC beta 0.9 from Mach 22, 1987](https://gcc.gnu.org/releases.html).

File: asmcomp/arm64/emit.mlp (hunk 2 above)
Comment (ghost) [reply asmcomp/arm64/emit.mlp:962]: I am trying to keep these changes consistent with the existing code style. Here this logic is easy to compare to `output_epilogue` and the `Lprologue` case in `emit_instr`, and this is intentional.

File: runtime/arm64.S (hunk 3 above)
Comment (ghost) [reply runtime/arm64.S:842]: I have considered using `C_ARG_#` defines in the new code, but unfortunately the rest of this file tends to never use them but use register names directly, so for now I want to remain consistent.

A further cleanup PR can be made later to change this.

That said, the use of `C_ARG_#` really only matters on amd64 where the code supports multiple ABIs and register names are bizarre for historical reasons. On all risc platforms, incqming argument registers are always rN, rN+1, rN+2, etc, so the need for symbolic names is much less important.

File: configure.ac (hunk 4 above)
Comment (ghost) [reply configure.ac]: Yes, and it's obviously not a C99 compiler, so it really doesn't matter to handle it. But I can undo that particular change if people prefer the old version - that change is indeed not relevant to arm64 TSan support.

File: runtime/arm64.S (hunk 5 above)
Comment (ghost) [reply runtime/arm64.S:1226]: No problem here. The `CFI_RESTORE_STATE` a few lines above matches the `CFI_REMEMBER_STATE` a few more line above, then `TSAN_EXIT_FUNCTION` keeps the stack state unchanged upon return, and reloading the possible return values is done from registers, so no stack impact either.

File: runtime/arm64.S (hunk 3 above)
Comment (ghost) [reply runtime/arm64.S:842]: Also note the general comments in the added code have been carried over from `amd64.S` - there are only a few arm64-specific new comments, those that explain the special +16 adjustment of stack pointers prior to invoking some tsan functions.

Comment (tmcgilchrist) [reply runtime/arm64.S:842]: I agree, personally I'd find it clearer to just use the ARM64 register names consistently in this file (in both the code and the comments sections. 
Maybe if there is general agreement that could be a cleanup PR later on?

Comment (ghost) [reply runtime/arm64.S:842]: You're right, it does not really make sense to keep `C_ARG_#` in the comments since the code does not make use of them. I'll adjust the comments.

**Review by gasche — APPROVED** [review]
The code looks reasonable, I trust the testing. I cannot vouch for the correctness of the new arm64 code, but I reviewed that there should be no change in behavior unless TSan is enabled. Approved!

**Comment by gasche** [comment]
@Octachron I would like to have this in 5.2, as TSan is a 5.2-only feature and we have a fair number of osx/arm64 users. Can we?

File: Changes [hunk 6]
```diff
@@ -200,6 +200,9 @@ Working version
   style that the compiler driver uses.
   (David Allsopp, review by Gabriel Scherer)
 
+- #12810: Port ThreadSanitizer support to Linux and macOS on arm64
+  (Miod Vallat, review by Tim McGilchrist)
```
Comment (gasche) [inline Changes:204]: +1

Comment (ghost) [reply Changes:204]: Added.

**Comment by Octachron** [comment]
Well, we already agreed that this should be part of the 5.2 milestone, so feel free to cherry-pick it.

**Comment by gasche** [comment]
Ah, sorry for the noise.
